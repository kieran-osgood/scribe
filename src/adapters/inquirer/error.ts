import { Data } from 'effect';

export class PromptError extends Data.TaggedClass('PromptError')<{
  readonly message: string;
  readonly cause?: unknown;
}> {
  override toString = () => this.message;
}
