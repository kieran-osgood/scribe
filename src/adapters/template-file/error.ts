import { Data } from 'effect';

export class TemplateFileError extends Data.TaggedClass('TemplateFileError')<{
  readonly error?: unknown;
}> {
  override toString(): string {
    return 'Writing to file failed, please report this.';
  }
}
