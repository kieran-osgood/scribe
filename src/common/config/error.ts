import { TreeFormatter } from '@effect/schema';
import { ParseError } from '@effect/schema/ParseResult';
import { Data } from 'effect';

export class CosmicConfigError extends Data.TaggedClass('CosmicConfigError')<{
  readonly error:
    | `[read config failed] ${string}`
    | 'invalid config'
    | 'Empty Config'
    | 'No template options found';
}> {
  override toString() {
    return `Config Read Error: ${this.error}`;
  }
}

export class ConfigParseError extends Data.TaggedClass('ConfigParseError')<{
  readonly parseError: ParseError;
  readonly path: string;
}> {
  override toString() {
    return `⚠️ Config parsing error: '${this.path}' 
 ${TreeFormatter.formatErrors(this.parseError.errors)}`;
  }
}
