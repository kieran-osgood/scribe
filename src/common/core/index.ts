// effect
export * as Boolean from '@effect/data/Boolean';
export * as Either from '@effect/data/Either';
export * from '@effect/data/Function';
export * as Option from '@effect/data/Option';
export * as Data from '@effect/data/Data';
// io
export * as Effect from '@effect/io/Effect';
export * as Runtime from '@effect/io/Runtime';
export * as Exit from '@effect/io/Exit';

//schema
export * as S from '@effect/schema/Schema';
export * as TF from '@effect/schema/TreeFormatter';

// export * as Sink from '@effect/stream/Sink';
// export * as Stream from '@effect/stream/Stream';

// core-ts
export * as IO from 'fp-ts/lib/IO.js';
export * as TE from 'fp-ts/lib/TaskEither.js';
export * as C from 'fp-ts/lib/Console.js';

// core-ts-std
export * as SStd from 'fp-ts-std/Struct';
