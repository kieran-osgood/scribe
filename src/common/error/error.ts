import { ZodError } from 'zod';

export function fmtError(error: unknown) {
  if (error instanceof ZodError) {
    return JSON.stringify(error.format());
  }

  if (error instanceof Error) {
    return error.message;
  }

  return String(error);
}
