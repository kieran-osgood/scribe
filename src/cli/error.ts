import { TaggedClass } from 'src/core';

export class CliError extends TaggedClass('CliError')<{
  readonly cause?: unknown;
}> {}
