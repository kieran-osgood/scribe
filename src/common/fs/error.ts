import { Data } from '@scribe/core';

export class ErrnoError extends Data.TaggedClass('ErrnoError')<{
  readonly error: NodeJS.ErrnoException;
}> {}

export class FileNotFoundError extends Data.TaggedClass(
  'fs.FileNotFoundError'
)<{
  readonly filePath: string;
}> {}
