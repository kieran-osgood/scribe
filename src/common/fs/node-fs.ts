import { Effect, pipe } from '@scribe/core';
import * as NFS from 'fs';
import { ErrnoError } from './error';
import { Abortable } from 'node:events';
import path from 'path';

export const writeFile = (
  pathName: string,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions
) =>
  Effect.async<never, ErrnoError, string>(resume => {
    NFS.writeFile(pathName, data, options, error => {
      if (error) {
        resume(Effect.fail(new ErrnoError({ error })));
      } else {
        resume(Effect.succeed(pathName));
      }
    });
  });

export const writeFileWithDir = (
  pathName: string,
  data: string | NodeJS.ArrayBufferView,
  options: NFS.WriteFileOptions
) =>
  pipe(
    mkdir(path.dirname(pathName), { recursive: true }),
    Effect.flatMap(() => writeFile(pathName, data, options))
  );

export const readFile = (
  path: NFS.PathOrFileDescriptor,
  options:
    | ({
        encoding?: BufferEncoding;
        flag?: string | undefined;
      } & Abortable)
    | undefined
    | null
) =>
  Effect.async<never, ErrnoError, string | Buffer>(resume => {
    NFS.readFile(path, options, (error, data) => {
      if (error) {
        resume(Effect.fail(new ErrnoError({ error })));
      } else {
        resume(Effect.succeed(data));
      }
    });
  });

export const mkdir = (
  path: NFS.PathLike,
  options: NFS.MakeDirectoryOptions & {
    recursive: true;
  }
) =>
  Effect.async<never, ErrnoError, string | undefined>(resume => {
    NFS.mkdir(path, options, (error, data) => {
      if (error) {
        resume(Effect.fail(new ErrnoError({ error })));
      } else {
        resume(Effect.succeed(data));
      }
    });
  });

export const stat = (
  filePath: string
): Effect.Effect<never, ErrnoError, NFS.Stats> =>
  Effect.async(resume =>
    NFS.stat(filePath, (error, stats) => {
      if (error) {
        resume(Effect.fail(new ErrnoError({ error })));
      } else {
        resume(Effect.succeed(stats));
      }
    })
  );

export const fileOrDirExists = (
  pathLike: string
): Effect.Effect<never, ErrnoError, boolean> =>
  pipe(
    stat(pathLike),
    Effect.map(_ => _.isFile() || _.isDirectory())
  );
