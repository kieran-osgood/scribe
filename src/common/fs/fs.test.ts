import { Effect, pipe } from '@scribe/core';
import { fs, vol } from 'memfs';
import { cwdAsJson } from '../../../configs/vite/setup-fs';
import * as FS from './node-fs';
import path from 'path';
import { ErrnoError } from './error';

const fileContents = 'super secret file';
beforeEach(() => {
  vi.restoreAllMocks();
});

describe('fs', () => {
  describe('readFile', () => {
    it('should read file to path', () =>
      pipe(
        Effect.gen(function* ($) {
          const filePath = './template1.ts';
          vol.writeFile(filePath, fileContents, err => {
            if (err) throw err;
          });

          const result = yield* $(FS.readFile(filePath, null));

          expect(String(result)).toBe(fileContents);
        }),
        Effect.runPromise
      ));

    it('throw error if file doesnt exist', () =>
      pipe(
        Effect.gen(function* ($) {
          const filePath = path.join(process.cwd(), './template2.ts');
          const result = yield* $(FS.readFile(filePath, null), Effect.flip);

          expect(result._tag).toBe('ErrnoError');
          expect(result).toBeInstanceOf(ErrnoError);
        }),
        Effect.runPromise
      ));
  });

  describe('writeFile', () => {
    it('should write file to path and read it back', () =>
      pipe(
        Effect.gen(function* ($) {
          const filePath = './template3.ts';

          expect(
            yield* $(FS.writeFile(filePath, fileContents, null)) //
          ).toBe('./template3.ts');
          expect(
            yield* $(FS.readFile(filePath, { encoding: 'utf8' })) //
          ).toEqual(fileContents);
        }),
        Effect.runPromise
      ));

    it('should write file to path and read it back', () =>
      pipe(
        Effect.gen(function* ($) {
          const filePath = '/some/nonexistent/path/template4.ts';

          const result = yield* $(
            pipe(FS.writeFile(filePath, fileContents, null), Effect.flip)
          );
          expect(result).toBeInstanceOf(ErrnoError);
        }),
        Effect.runPromise
      ));
  });

  describe('mkdir', () => {
    it('creates recursively ', () =>
      pipe(
        Effect.gen(function* ($) {
          yield* $(FS.mkdir('./some/nested/pathway', { recursive: true }));
          expect(cwdAsJson()).toMatchSnapshot();

          const dirExists = fs.existsSync('./some/nested/pathway');
          expect(dirExists).toEqual(true);

          yield* $(FS.mkdir('./tmpl/a/b', { recursive: true }));
          expect(cwdAsJson()).toMatchSnapshot();
        }),
        Effect.runPromise
      ));

    it('returns undefined if dir already exists', () =>
      pipe(
        Effect.gen(function* ($) {
          const filePath = './some/path';
          vol.mkdirSync(filePath, { recursive: true });
          expect(cwdAsJson()).toMatchSnapshot();

          const result = yield* $(FS.mkdir(filePath, { recursive: true }));
          expect(result).toBe(undefined);
          expect(cwdAsJson()).toMatchSnapshot();
        }),
        Effect.runPromise
      ));
  });

  describe('writeFileWithDir', () => {
    it('should write file to path and read it back', () =>
      pipe(
        Effect.gen(function* ($) {
          const filePath = '/path/to/template5.ts';

          const result = yield $(
            FS.writeFileWithDir(filePath, fileContents, null)
          );
          expect(result).toBe(filePath);

          const readResult = yield* $(
            FS.readFile(filePath, { encoding: 'utf8' })
          );
          expect(readResult).toEqual(fileContents);
        }),
        Effect.runPromise
      ));
  });
});
