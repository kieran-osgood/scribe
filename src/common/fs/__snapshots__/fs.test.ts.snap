// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fs > mkdir > creates recursively  1`] = `
{
  "some/nested/pathway": null,
}
`;

exports[`fs > mkdir > creates recursively  2`] = `
{
  "some/nested/pathway": null,
  "tmpl/a/b": null,
}
`;

exports[`fs > mkdir > returns undefined if dir already exists 1`] = `
{
  "some/path": null,
}
`;

exports[`fs > mkdir > returns undefined if dir already exists 2`] = `
{
  "some/path": null,
}
`;
